### ISSUE TITLE
Memory spike at opening service

### ISSUE THREAD

**Issue by KeyTry** [cboulay/PSMoveService#322]
I found this issue posted on two other threads, [this one](https://github.com/cboulay/PSMoveService/issues/89) and [this one](https://github.com/cboulay/PSMoveService/issues/127). However, there doesn't seem to be a solution that works for me. I was using the program fine yesterday. I'm curious if someone has managed to identify the issue or another solution that doesn't involve deleting the boost_interprocess files.

Basically, whenever I try to use the PSMoveService.exe, I have a huge memory spike. The log is rather simple:

```
[2017-01-05 17:05:53.422]: main - Starting PSMoveService
[2017-01-05 17:05:53.446]: bluetooth_get_host_address - Found a bluetooth radio
[2017-01-05 17:05:53.447]: bluetooth_get_host_address - Retrieved radio info
[2017-01-05 17:05:53.488]: PS3EyeTracker::open - Opening PS3EyeTracker(USB\VID_1415&PID_2000\7\0, camera_index=0)
[2017-01-05 17:06:14.263]: SharedMemory::initialize() - Allocating shared memory: tracker_view_0
[2017-01-05 17:06:18.113]: EXCEPTION - PSMoveService - boost::interprocess::intermodule_singleton initialization failed
[2017-01-05 17:06:18.869]: ClientConnection::stop - Stopping client connection id 0
[2017-01-05 17:06:21.738]: SharedMemory::dispose - Failed to free shared memory: tracker_view_0
```

I'm not using CCleaner or any other registry cleaner.

I believe a few updates for Windows and Visual Studio might have affected it somehow. However, I used the program yesterday right after these updates. It only stopped working this morning. In any case, these are my updates.

![updaets](https://cloud.githubusercontent.com/assets/15346451/21701720/fc6aafac-d36c-11e6-9787-e773c3fe8d18.PNG)

**Comment by HipsterSloth** [comment]
I just found this issue report in the boost project issue tracker:

https://svn.boost.org/trac/boost/ticket/12137

It looks like boost::interprocess::intermodule_singleton looks through the windows event logs to find [event 6005](https://www.microsoft.com/technet/support/ee/transform.aspx?ProdName=Windows+Operating+System&ProdVer=5.2&EvtID=6005&EvtSrc=EventLog) which it then uses to compute the up-time of the system. If your system has been running for a while it will clean out old events. This is not an ideal way to get this information, but I don't want to make a custom build of boost (since it would add significant build times to people who build psmoveservice locally). However, it sounds like you can work around this simply by rebooting your machine.

TL;DR Try rebooting your PC and see if that fixes it.

**Comment by KeyTry** [comment]
> TL;DR Try rebooting your PC and see if that fixes it.
Oh, yes, that's one of the first things I tried. I've rebooted, deleted the boost_inteprocess files, and reinstalled the tracker drivers.

**Comment by HipsterSloth** [comment]
Hmm. Can you verify that your system has emitted the Windows event 6005 using "Windows Event Viewer"? 

Here is what my event log looks like (Windows 10) when I filter on event 6005:
![image](https://cloud.githubusercontent.com/assets/16793853/21702994/d45e8cc6-d363-11e6-8f71-726586301cb0.png)

**Comment by KeyTry** [comment]
> Hmm. Can you verify that your system has emitted the Windows event 6005 using "Windows Event Viewer"?

Yes, I have quite a few recent 6005 events (The OS is in spanish, but it should be about the same)

![events](https://cloud.githubusercontent.com/assets/15346451/21703367/1c3da33e-d377-11e6-88e6-96d6bbe1ebb6.PNG)

**Comment by KeyTry** [comment]
I tried running the service a few more times, and it seems event 6005 isn't being created most of the time. I'm unsure why I have a few recent events, but most do not create it.

**Comment by HipsterSloth** [comment]
> I tried running the service a few more times, and it seems event 6005 isn't being created most of the time. I'm unsure why I have a few recent events, but most do not create it.

Event 6005 should only be created on reboot of your PC, not on PSMoveService restart.

In any event, if event 6005 exists, then that likely isn't the issue. I think something else might be preventing you from creating the boost interprocess file. 

Do you have the following registry key:
"HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Explorer\Shell Folders"?

If so, is the "Common AppData" entry "c:\ProgramData"?

Is there a "C:\ProgramData\boost_interprocess" folder?

If yes to all of those questions, then the next thing to check would be to run ProcMon and see what happens when PSMoveService tries to access the boost_interprocess folder.

**Comment by KeyTry** [comment]
> If yes to all of those questions, then the next thing to check would be to run ProcMon and see what happens when PSMoveService tries to access the boost_interprocess folder.

Yes to all those. This is what I'm getting using this filter on ProcMon.

![procmon1](https://cloud.githubusercontent.com/assets/15346451/21704332/3329dc0a-d37e-11e6-9622-7ed21f752009.PNG)

**Comment by KeyTry** [comment]
According to [this thread](https://github.com/networkMe/missinghud2/issues/8) on a similar error, there should a numbered folder inside boost_interprocess. I can't find such a folder. The OP says they had it fixed after rebooting the machine, but that doesn't seem to do it for me.

**Comment by HipsterSloth** [comment]
So the "Name Collision" error on the boost_interprocess folder would imply that it's trying to create a new temp folder with a unique numbered folder, but it's already finding an existing numbered folder with that same number. However that's confusing since the boost_interprocess folder is empty on youe machine. 

Were there any other entries in ProcMon for PSMoveService.exe if you removed the "c:\ProgramData\boost_interprocess" filter?

**Comment by KeyTry** [comment]
> Were there any other entries in ProcMon for PSMoveService.exe if you removed the "c:\ProgramData\boost_interprocess" filter?

Oh, yes. 5201 to be precise.

![procmon2](https://cloud.githubusercontent.com/assets/15346451/21705415/74bcd1ca-d385-11e6-8349-38b4e56fcbfb.PNG)

In case it's of any interest, [here's](https://mega.nz/#!W003CIBD!sVMyADkktPXbgq-xZrI_kkD3Dyyj4CHnfOJigqqL2N0) the PML with the PSMove data.

**Comment by HipsterSloth** [comment]
So I looked over the full PML log. I think the "Name Collision" event is actually normal. It looks like boost will always try to create the boost_interprocess directory. If it already exists it'll get the "Name Collision" event. I didn't see anything else out of the ordinary in the event log.

After reading a bit more about how the boost::interprocess handles [directory management](http://www.boost.org/doc/libs/1_59_0/doc/html/interprocess/acknowledgements_notes.html#interprocess.acknowledgements_notes.notes_linux.notes_linux_shm_folder) it looks like you can override where the memory map folder goes. Since both PSMoveService and PSMoveConfigTool (the two users of shared memory for the video feeds) live in the same directory I made a custom build that puts the shared memory files in the same directory as PSMoveService under "shared_mem":

[PSMoveService-v0.9alpha6_0_8_custom.zip](https://github.com/cboulay/PSMoveService/files/689112/PSMoveService-v0.9alpha6_0_8_custom.zip)

Let me know if this custom build works for you. You should see a "shared_mem" folder show up with "tracker_view_X" files in it (one for each tracker that opens) when you open PSMoveService.

**Comment by KeyTry** [comment]
Hey, thank you! the custom build seems to work just fine. I deeply appreciate the time and effort.

I'll post on the Google Groups board if there's another issue on the way, but so far everything seems alright.

**Comment by HipsterSloth** [comment]
Cool! Glad to hear that's working for you. I'll put out a build late tonight with this fix (as well as one for controller orientation jitter and incorrect controller models in steam vr) and then close this issue once the official build is up. 

Thanks for all you help with the debugging!

**Comment by HipsterSloth** [comment]
Fixed in latest release: https://github.com/cboulay/PSMoveService/releases/tag/v0.9-alpha6.0.8
